acceptance/tests/resource/exec/should_run_command_in_cwd.rb

```diff
@@ -2,117 +2,180 @@
+          #
+          # To get around this for MacOS platforms we simply use the new user's homedir as the 'tmpdir' and
+          # put the noaccess dir there.
+          on(agent, "mkdir /Users/#{username}/noaccess_test && cp #{tmpdir_noaccess}/noaccess.txt /Users/#{username}/noaccess_test && chmod -R 600 /Users/#{username}/noaccess_test")
+          tmpdir_noaccess = "/Users/#{username}/noaccess_test"
+        end
+        # remove permissions for all other users other than root, which should force puppet to fail when running as another user
+        on(agent, "chmod -R 600 #{tmpdir_noaccess}")
       end
+    end
 
-      manifest = exec_resource_manifest.call(command: command, cwd: cwd, onlyif: onlyif)
-
-      # debug: true prints the output of the onlyif command
-      apply_manifest_on(agent, manifest, debug: true) do |result|
-        assert_match(to_regex(cwd), result.stdout, 'The Exec resource runs a "check" command in the PWD instead of the CWD.')
+    step "Errors if the user does not have access to the specified CWD" do
+      manifest_path = agent.tmpfile('apply_manifest.pp')
+      create_remote_file(agent, manifest_path, exec_resource_manifest("#{cat} noaccess.txt", {:cwd => tmpdir_noaccess, :path => path}))
+      if agent.platform =~ /windows/
+        on(agent, "cmd.exe /c \"puppet apply #{manifest_path} --detailed-exitcodes\"", :acceptable_exit_codes => [4]) do |result|
+          assert_equal(4, result.exit_code, "Exec manifest still executed inside restricted directory", )
+        end
+      elsif agent.platform =~ /osx/
+        # on MacOS we need to copy the manifest to run to the user's home dir and give the user ownership. otherwise puppet won't run on it.
+        on(agent, "cp #{manifest_path} /Users/#{username}/noaccess_manifest.pp && chown #{username}:80 /Users/#{username}/noaccess_manifest.pp")
+        on(agent, "su - #{username} -c \"/opt/puppetlabs/bin/puppet apply /Users/#{username}/noaccess_manifest.pp --detailed-exitcodes\"", :acceptable_exit_codes => [4]) do |result|
+          assert_equal(4, result.exit_code, "Exec manifest still executed inside restricted directory")
+        end
+      else
+        on(agent, "chown #{username} #{manifest_path}")
+        if agent.platform =~ /solaris|aix/
+          on(agent, "su - #{username} -c \"/opt/puppetlabs/bin/puppet apply #{manifest_path} --detailed-exitcodes\"", :acceptable_exit_codes => [4]) do |result|
+            assert_equal(4, result.exit_code, "Exec manifest still executed inside restricted directory")
+          end
+        else
+          on(agent, "su #{username} -c \"/opt/puppetlabs/bin/puppet apply #{manifest_path} --detailed-exitcodes\"", :acceptable_exit_codes => [4]) do |result|
+            assert_equal(4, result.exit_code, "Exec manifest still executed inside restricted directory")
+          end
+        end
       end
     end
+
+    step "remove test user once testing with it is done" do
```

Comment:
It'd be better to put this in a teardown block at the point where you're creating the user.